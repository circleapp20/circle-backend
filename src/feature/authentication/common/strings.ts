import { VERIFICATION_CODE_EMAIL_TEMPLATE } from 'feature/authentication/common/templates';

export const getVerificationEmailText = (code: string, user = 'User') => {
	const template = VERIFICATION_CODE_EMAIL_TEMPLATE;
	return template.replace('{verificationCode}', code).replace('{user}', user);
};
