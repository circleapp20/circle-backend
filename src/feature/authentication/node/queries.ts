import { Users } from 'shared/common/schema/users';
import { EntityManager } from 'typeorm';

export const countMatchingIdAndCodeQuery = (
	manager: EntityManager,
	id: string,
	verificationCode: string
) => {
	return manager
		.createQueryBuilder(Users, 'u')
		.where('u.id = :id', { id })
		.andWhere('u.verificationCode = :code', { code: verificationCode })
		.getCount();
};

export const updateUserVerificationCodeQuery = (
	manager: EntityManager,
	values: { id: string; verificationCode: string }
) => {
	const { id, verificationCode } = values;
	return manager
		.createQueryBuilder(Users, 'u')
		.update()
		.set({ verificationCode })
		.where('u.id = :id', { id })
		.execute();
};
