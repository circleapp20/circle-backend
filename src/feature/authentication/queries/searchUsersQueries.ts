import { Users } from 'base/common/schema/users';
import { EntityManager } from 'typeorm';

export const countMatchingIdAndCodeQuery = (
	manager: EntityManager,
	id: string,
	verificationCode: string
) => {
	return manager
		.createQueryBuilder(Users, 'u')
		.where('u.id = :id', { id })
		.andWhere('u.verificationCode = :code', { code: verificationCode })
		.getCount();
};
