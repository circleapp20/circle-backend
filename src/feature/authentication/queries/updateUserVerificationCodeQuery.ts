import { Users } from 'base/common/schema/users';
import { EntityManager } from 'typeorm';

export const updateUserVerificationCodeQuery = (
	manager: EntityManager,
	values: { id: string; verificationCode: string }
) => {
	const { id, verificationCode } = values;
	return manager
		.createQueryBuilder(Users, 'u')
		.update()
		.set({ verificationCode })
		.where('id = :id', { id })
		.execute();
};
