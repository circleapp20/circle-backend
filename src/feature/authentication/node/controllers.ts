import { Response } from 'express';
import {
	checkUserVerificationCode,
	createUserProfileWithDefaultValues,
	getUserAccountWithCredentials,
	getUserProfileById,
	verifyUserLoginCredentials
} from 'feature/authentication/node/dataService';
import { sendVerificationCodeByMedia } from 'feature/authentication/node/messaging';
import { getResponseData } from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { IAuthUser, IRequest } from 'shared/types';

export const verifyUserCredentials = async (req: IRequest, res: Response) => {
	const user = await createUserProfileWithDefaultValues(req.body.data);

	sendVerificationCodeByMedia({
		media: user.email ? 'email' : 'phoneNumber',
		verificationCode: user.verificationCode,
		email: user.email,
		phoneNumber: user.phoneNumber
	});

	const responseData = getResponseData(user);
	res.status(Constants.status.CREATED).json(responseData);
};

export const verifyUserVerificationCode = async (req: IRequest, res: Response) => {
	const { id }: IAuthUser = req.user;
	const data = Object.assign({}, req.body.data, { id });
	const status = await checkUserVerificationCode(data);
	const responseData = getResponseData(status);
	res.status(Constants.status.CREATED).json(responseData);
};

export const verifyUserLogin = async (req: IRequest, res: Response) => {
	const user = await verifyUserLoginCredentials(req.body.data);
	const responseData = getResponseData(user);
	res.status(Constants.status.CREATED).json(responseData);
};

export const resendUserVerificationCode = async (req: IRequest, res: Response) => {
	const { media }: any = req.query;
	const user = await getUserProfileById(req.user.id);

	sendVerificationCodeByMedia({
		media: media ? media : user.email ? 'email' : 'phoneNumber',
		verificationCode: user.verificationCode,
		email: user.email,
		phoneNumber: user.phoneNumber
	});

	const responseData = getResponseData(true);
	res.status(Constants.status.CREATED).json(responseData);
};

/**
 * the controller verifies the user's credentials in the database
 * and sends the verification code to the user by email or password
 * should the user have both email and password, a response of message
 * is return to the user to request which media to send the verification
 * code to either email or phone number
 */
export const verifyUserCredentialsForPasswordReset = async (req: IRequest, res: Response) => {
	const profile = await getUserAccountWithCredentials(req.body.data);

	const { email, verificationCode, phoneNumber } = profile;
	let message = 'Verification code cannot be sent. User has both email and phone number';

	sendVerificationCodeByMedia({
		media: phoneNumber ? 'phoneNumber' : 'email',
		verificationCode,
		email: email && !phoneNumber ? email : '',
		phoneNumber: !email && phoneNumber ? phoneNumber : ''
	});

	if (email && !phoneNumber) {
		message = "Verification code sent to user's email";
	} else if (!email && phoneNumber) {
		message = "Verification code sent to user's phone number";
	}

	const responseData = getResponseData({ user: profile, message });
	res.status(Constants.status.CREATED).json(responseData);
};
