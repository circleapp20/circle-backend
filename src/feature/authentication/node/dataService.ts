import bcryptjs from 'bcryptjs';
import {
	countMatchingIdAndCodeQuery,
	updateUserVerificationCodeQuery
} from 'feature/authentication/node/queries';
import { getBadRequestError } from 'shared/common/errors';
import { generateCodeFromNumber } from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { runInsertQuery, runInTransaction, runQuery } from 'shared/node/database';
import {
	addUserProfileQuery,
	getUserByCredentialsQuery,
	getUserByIdQuery
} from 'shared/node/queries';
import { getSignedAuthToken } from 'shared/node/validation';
import { IAddUserProfile } from 'shared/types';
import { EntityManager } from 'typeorm';

export const addUserTransaction = (email = '', phoneNumber = '') => {
	return async (manager: EntityManager) => {
		// generate a verification code for the user
		const verificationCode = generateCodeFromNumber();

		// construct a default user info
		const profile: IAddUserProfile = {
			username: '',
			password: '',
			dob: new Date(),
			image: '',
			biography: '',
			email: email || '',
			phoneNumber: phoneNumber || '',
			isEmailVerified: false,
			verificationCode,
			roles: [Constants.privileges.USER],
			name: ''
		};

		const [user] = await runInsertQuery(addUserProfileQuery, [profile], manager);

		return runQuery(getUserByIdQuery, [user.id], manager);
	};
};

export const createUserProfileWithDefaultValues = async (data: {
	email?: string;
	phoneNumber?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (user) throw getBadRequestError('User already exists');

	// save user profile into the database
	const userProfile = await runInTransaction(addUserTransaction(data.email, data.phoneNumber));

	// add token to user profile
	const token = getSignedAuthToken({ id: userProfile!.id, roles: [Constants.privileges.USER] });
	const profile = Object.assign({}, userProfile, { token });

	return profile;
};

export const checkUserVerificationCode = async (data: { id: string; verificationCode: string }) => {
	const { id, verificationCode } = data;

	if (!id || !verificationCode) {
		throw getBadRequestError('user id and verification code are required');
	}

	const count = await runQuery(countMatchingIdAndCodeQuery, [id, verificationCode]);
	if (!count) throw getBadRequestError('invalid verification code');

	return true;
};

export const verifyUserLoginCredentials = async (data: {
	email?: string;
	phoneNumber?: string;
	username?: string;
	password: string;
}) => {
	const { phoneNumber, password, email, username } = data;
	if (!phoneNumber && !email && !username) {
		throw getBadRequestError('phoneNumber, email or username is required');
	}
	if (!password) throw getBadRequestError('password is required');

	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (!user) throw getBadRequestError('Invalid user account');

	if (!bcryptjs.compareSync(password, user.password)) {
		throw getBadRequestError('Invalid user account');
	}

	const { password: hashedPassword, ...other } = user;

	// create user token
	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token });

	return profile;
};

export const getUserProfileById = async (id: string) => {
	const user = await runQuery(getUserByIdQuery, [id]);
	if (!user) throw getBadRequestError('Invalid user account');
	return user;
};

/**
 * the service function validates the user's credentials in the
 * database to check if the user exists
 * if the user does exists, a new verification code is generated
 * and stored in the database
 */
export const getUserAccountWithCredentials = async (args: {
	email?: string;
	phoneNumber?: string;
	username?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [args]);
	if (!user) throw getBadRequestError('Invalid user account');

	// generate new verification code
	const verificationCode = generateCodeFromNumber();

	// update user profile
	await runQuery(updateUserVerificationCodeQuery, [{ id: user.id, verificationCode }]);

	const { password, ...other } = user;

	const token = getSignedAuthToken({ id: user.id, roles: user.roles });
	const profile = Object.assign({}, other, { token, verificationCode });

	return profile;
};
