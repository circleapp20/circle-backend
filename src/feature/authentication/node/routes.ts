import * as schema from 'feature/authentication/common/dataSchema';
import * as controller from 'feature/authentication/node/controllers';
import { Constants } from 'shared/constants';
import { IApiRoute } from 'shared/types';

export const authRoutes: IApiRoute[] = [
	{
		path: '/auth/sign-up',
		method: 'post',
		controller: controller.verifyUserCredentials,
		schema: schema.VerifyUserCredentialsSchema
	},
	{
		path: '/auth/code',
		method: 'post',
		controller: controller.verifyUserVerificationCode,
		privileges: [Constants.privileges.SUPER_ADMIN, Constants.privileges.USER],
		schema: schema.VerifyVerificationCodeSchema
	},
	{
		path: '/auth/sign-in',
		method: 'post',
		controller: controller.verifyUserLogin,
		schema: schema.VerifyUserLoginSchema
	},
	{
		path: '/auth/code/resend',
		method: 'post',
		controller: controller.resendUserVerificationCode,
		privileges: [Constants.privileges.USER]
	},
	{
		path: '/auth/credentials',
		method: 'post',
		controller: controller.verifyUserCredentialsForPasswordReset
	}
];
