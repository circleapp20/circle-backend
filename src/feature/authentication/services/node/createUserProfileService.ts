import { SERVER } from 'base/config/server';
import { IAddUserProfile } from 'base/types';
import { getBadRequestError } from 'base/utils/errors/node/badRequestError';
import { generateCodeFromNumber } from 'base/utils/node/codeGenerator';
import { getSignedAuthToken } from 'base/utils/server/validation';
import { runInsertQuery, runInTransaction, runQuery } from 'core/database/queryRunners';
import { decryptData, encryptData } from 'core/encryption/node/encryption';
import {
	addUserProfileQuery,
	getUserByCredentialsQuery,
	getUserByIdQuery
} from 'core/queries/userQueries';
import { EntityManager } from 'typeorm';

export const addUserTransaction = (email = '', phoneNumber = '') => {
	return async (manager: EntityManager) => {
		// generate a verification code for the user
		const verificationCode = generateCodeFromNumber();
		const encryptedVerificationCode = encryptData({ text: verificationCode });

		// construct a default user info
		const profile: IAddUserProfile = {
			username: '',
			password: '',
			dob: new Date(),
			image: '',
			biography: '',
			email: email || '',
			phoneNumber: phoneNumber || '',
			isEmailVerified: false,
			verificationCode: encryptedVerificationCode,
			roles: [SERVER.privileges.USER],
			name: ''
		};

		const [user] = await runInsertQuery(addUserProfileQuery, [profile], manager);

		return runQuery(getUserByIdQuery, [user.id], manager);
	};
};

export const createUserProfileWithDefaultValues = async (data: {
	email?: string;
	phoneNumber?: string;
}) => {
	const user = await runQuery(getUserByCredentialsQuery, [data]);
	if (user) throw getBadRequestError('User already exists');

	// save user profile into the database
	const userProfile = await runInTransaction(addUserTransaction(data.email, data.phoneNumber));
	const verificationCode = decryptData({ encryptedText: userProfile!.verificationCode });

	// add token to user profile
	const token = getSignedAuthToken({
		id: userProfile!.id,
		roles: [SERVER.privileges.USER]
	});
	const profile = Object.assign({}, userProfile, { token, verificationCode });

	return profile;
};
