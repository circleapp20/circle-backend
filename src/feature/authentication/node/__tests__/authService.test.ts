import bcryptjs from 'bcryptjs';
import * as dataService from 'feature/authentication/node/authService';
import { createUserFixture } from 'fixtures/users';
import * as utils from 'shared/common/utilities';
import { Constants } from 'shared/constants';
import { runQuery } from 'shared/node/database';
import * as encryption from 'shared/node/encryption';
import * as sharedQueries from 'shared/node/queries';
import { entityManager } from 'shared/testUtils/node/entityManager';

jest.mock('shared/node/database');
jest.mock('shared/common/schema/users');
jest.mock('shared/common/schema/fellows');
jest.mock('shared/common/schema/locations');
jest.mock('bcryptjs', () => ({
	hashSync: jest.fn().mockReturnValue('$$20yy39nv93n932n92093nf92'),
	compareSync: jest.fn()
}));

beforeEach(() => jest.clearAllMocks());

describe('#addUserTransaction', () => {
	const addUserQuerySpy = jest.spyOn(sharedQueries, 'addUserProfileQuery');
	let createUserTransaction: (manager: any) => Promise<any>;

	beforeEach(() => {
		entityManager.execute.mockReturnValueOnce({
			generateMaps: [{ id: 'x7i9-3l-n3k4-3i8bi2' }]
		});
		createUserTransaction = dataService.addUserTransaction('test@test.com');
	});

	test('should create user with email', async () => {
		await createUserTransaction(entityManager);
		expect(addUserQuerySpy).toHaveBeenCalled();
	});

	test('should create user with phoneNumber', async () => {
		const transaction = dataService.addUserTransaction('', '+2339874563');
		await transaction(entityManager);
		expect(addUserQuerySpy).toHaveBeenCalled();
	});

	test('should generate random code for verification', async () => {
		const spy = jest.spyOn(utils, 'generateCodeFromNumber');
		await createUserTransaction(entityManager);
		expect(spy).toHaveBeenCalledWith();
	});

	test('should encrypt verification code', async () => {
		const spy = jest.spyOn(encryption, 'encryptData');
		await createUserTransaction(entityManager);
		expect(spy).toHaveBeenCalledWith(expect.objectContaining({ text: expect.any(String) }));
	});

	test('should add the user with the default profile', async () => {
		await createUserTransaction(entityManager);
		expect(addUserQuerySpy).toHaveBeenCalledWith(
			entityManager,
			expect.objectContaining({
				username: '',
				password: '',
				email: expect.stringMatching('test@test.com'),
				phoneNumber: '',
				roles: expect.arrayContaining([Constants.privileges.USER]),
				verificationCode: expect.any(String)
			})
		);
	});

	test('should return the added user', async () => {
		const getUserQuerySpy = jest.spyOn(sharedQueries, 'getUserByIdQuery');
		await createUserTransaction(entityManager);
		expect(getUserQuerySpy).toHaveBeenCalledWith(entityManager, 'x7i9-3l-n3k4-3i8bi2');
	});

	test('should return the new user profile', async () => {
		entityManager.getOne.mockReturnValueOnce({
			id: 'x7i9-3l-n3k4-3i8bi2',
			verificationCode: '$$20yy39nv93n932n92093nf92',
			phoneNumber: '',
			email: 'test@test.com',
			roles: ['user']
		});
		const results = await createUserTransaction(entityManager);
		expect(results).toEqual(
			expect.objectContaining({
				id: expect.any(String),
				email: expect.stringMatching('test@test.com'),
				phoneNumber: '',
				roles: expect.arrayContaining([Constants.privileges.USER]),
				verificationCode: expect.any(String)
			})
		);
	});
});

describe('#checkUserVerificationCode', () => {
	test('should return true if code and id matches', async () => {
		const user = createUserFixture();
		entityManager.getOne.mockReturnValueOnce({
			...user,
			verificationCode: encryption.encryptData({ text: '345287' })
		});
		const data = { id: user.id, verificationCode: '345287' };
		const results = await dataService.checkUserVerificationCode(data);
		expect(results).toBeTruthy();
	});

	test('should throw if id is empty', () => {
		const data = { id: '', verificationCode: '3452' };
		expect(dataService.checkUserVerificationCode(data)).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw if verificationCode is empty', () => {
		const data = { id: 'x7i9-3l-n3k4-3i8bi2', verificationCode: '' };
		expect(dataService.checkUserVerificationCode(data)).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw if user is null', () => {
		entityManager.getOne.mockReturnValueOnce(null);
		const data = { id: 'x7i9-3l-n3k4-3i8bi2', verificationCode: '345287' };
		expect(dataService.checkUserVerificationCode(data)).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should decrypt stored encrypted verification code', async () => {
		const user = createUserFixture();
		entityManager.getOne.mockReturnValueOnce({
			...user,
			verificationCode: encryption.encryptData({ text: '345287' })
		});
		const spy = jest.spyOn(encryption, 'decryptData');
		const data = { id: 'x7i9-3l-n3k4-3i8bi2', verificationCode: '345287' };
		await dataService.checkUserVerificationCode(data);
		expect(spy).toHaveBeenCalledWith(
			expect.objectContaining({ encryptedText: expect.any(String) })
		);
	});
});

describe('#verifyUserLoginCredentials', () => {
	test('should throw error if username, email and phoneNumber is null', (done) => {
		dataService
			.verifyUserLoginCredentials({
				password: 'jellybean',
				email: '',
				phoneNumber: '',
				username: ''
			})
			.catch((error) => {
				expect(error.message).toBe('phoneNumber, email or username is required');
				done();
			});
	});

	test('should throw if there is no password', (done) => {
		dataService
			.verifyUserLoginCredentials({
				password: '',
				email: 'test@test.com',
				phoneNumber: '',
				username: ''
			})
			.catch((error) => {
				expect(error.message).toBe('password is required');
				done();
			});
	});

	test('should throw error if user is not found', (done) => {
		(runQuery.mockResolvedValueOnce as any)(null);
		dataService
			.verifyUserLoginCredentials({
				password: 'testing',
				email: 'test@test.com',
				phoneNumber: '',
				username: 'tester'
			})
			.catch((error) => {
				expect(error.message).toBe('Invalid user account');
				done();
			});
	});

	test('should throw error if password does not match', (done) => {
		(<any>bcryptjs).compareSync.mockReturnValueOnce(false);
		(runQuery.mockResolvedValueOnce as any)({ password: '$9j9h48v3ge92' });
		dataService
			.verifyUserLoginCredentials({
				password: 'testing',
				email: 'test@test.com',
				phoneNumber: '',
				username: 'tester'
			})
			.catch((error) => {
				expect(error.message).toBe('Invalid user account');
				done();
			});
	});

	test('should return user details with token', (done) => {
		(<any>bcryptjs).compareSync.mockReturnValueOnce(true);
		(runQuery.mockResolvedValueOnce as any)({
			password: '$9j9h48v3ge92',
			id: 'x7i9-3l-n3k4-3i8bi2'
		});
		dataService
			.verifyUserLoginCredentials({
				password: 'testing',
				email: 'test@test.com',
				phoneNumber: '',
				username: 'tester'
			})
			.then((user) => {
				expect(user.id).toBe('x7i9-3l-n3k4-3i8bi2');
				expect(user.token).toBeDefined();
				done();
			});
	});
});
describe('#getUserProfileById', () => {
	test('should call getUserByIdQuery with the user id', async () => {
		const user = createUserFixture();
		const verificationCode = encryption.encryptData({ text: '783465' });
		entityManager.getOne.mockReturnValueOnce(Object.assign(user, { verificationCode }));
		await dataService.getUserProfileById(user.id);
		expect(runQuery).toHaveBeenCalledWith(
			expect.any(Function),
			expect.arrayContaining([user.id])
		);
	});

	test('should throw error if user does not exists', () => {
		entityManager.getOne.mockReturnValueOnce(null);
		expect(
			dataService.getUserProfileById('x39-39ng39-nf39')
		).rejects.toThrowErrorMatchingSnapshot();
	});
});

describe('#getUserAccountWithCredentials', () => {
	const userProfile = createUserFixture();
	const credentials = {
		username: userProfile.username,
		phoneNumber: userProfile.phoneNumber,
		email: userProfile.email
	};

	beforeAll(() => {
		const verificationCode = encryption.encryptData({ text: '123456' });
		Object.assign(userProfile, { verificationCode });
	});

	test('should throw if user is not found', () => {
		entityManager.getOne.mockReturnValueOnce(null);
		expect(
			dataService.getUserAccountWithCredentials(credentials)
		).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should add token to the return object', async () => {
		entityManager.getOne.mockReturnValueOnce(userProfile);
		const user = await dataService.getUserAccountWithCredentials(credentials);
		expect(user.token).toEqual(expect.any(String));
	});

	test('should not return the password', async () => {
		entityManager.getOne.mockReturnValueOnce(userProfile);
		const user = await dataService.getUserAccountWithCredentials(credentials);
		expect(user).not.toHaveProperty('password');
	});

	test('should update user with new verification code', async () => {
		entityManager.getOne.mockReturnValueOnce(userProfile);
		await dataService.getUserAccountWithCredentials(credentials);
		expect(runQuery).toHaveBeenNthCalledWith(
			2,
			expect.any(Function),
			expect.arrayContaining([
				expect.objectContaining({
					id: expect.any(String),
					verificationCode: expect.any(String)
				})
			])
		);
	});

	test('should create new verification code', async () => {
		entityManager.getOne.mockReturnValueOnce(userProfile);
		const spy = jest.spyOn(utils, 'generateCodeFromNumber');
		await dataService.getUserAccountWithCredentials(credentials);
		expect(spy).toHaveReturnedWith(expect.any(String));
	});

	test('should encrypt the verification code', async () => {
		entityManager.getOne.mockReturnValueOnce(userProfile);
		const spy = jest.spyOn(utils, 'generateCodeFromNumber');
		spy.mockReturnValueOnce('678902');
		await dataService.getUserAccountWithCredentials(credentials);
		expect(runQuery).toHaveBeenNthCalledWith(
			2,
			expect.any(Function),
			expect.arrayContaining([
				{ id: userProfile.id, verificationCode: expect.not.stringMatching('678902') }
			])
		);
	});
});

describe('#createUserProfileWithDefaultValues', () => {
	const profile = createUserFixture();

	test('should add user if email does not exists', async () => {
		entityManager.getOne.mockReturnValueOnce(null).mockReturnValueOnce(profile);
		entityManager.execute.mockReturnValueOnce({ generateMaps: [profile] });
		const user = await dataService.createUserProfileWithDefaultValues({
			email: profile.email
		});
		expect(user).toEqual(
			expect.objectContaining({
				...profile,
				token: expect.any(String)
			})
		);
	});

	test('should create new user if phoneNumber does not exists', async () => {
		entityManager.getOne.mockReturnValueOnce(null).mockReturnValueOnce(profile);
		entityManager.execute.mockReturnValueOnce({ generateMaps: [profile] });
		const user = await dataService.createUserProfileWithDefaultValues({
			phoneNumber: profile.phoneNumber
		});
		expect(user).toEqual(
			expect.objectContaining({
				...profile,
				token: expect.any(String)
			})
		);
	});

	test('should throw an error if email already exists', () => {
		entityManager.getOne.mockReturnValueOnce(profile);
		expect(
			dataService.createUserProfileWithDefaultValues({ email: 'test@test.com' })
		).rejects.toThrowErrorMatchingSnapshot();
	});

	test('should throw bad request error if phone number already exists', () => {
		entityManager.getOne.mockReturnValueOnce(profile);
		expect(
			dataService.createUserProfileWithDefaultValues({ email: '+2339954853' })
		).rejects.toThrowErrorMatchingSnapshot();
	});
});
