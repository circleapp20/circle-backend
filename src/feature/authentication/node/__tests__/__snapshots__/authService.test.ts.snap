// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#checkUserVerificationCode should throw if id is empty 1`] = `"user id and verification code are required"`;

exports[`#checkUserVerificationCode should throw if user is null 1`] = `"Invalid user account"`;

exports[`#checkUserVerificationCode should throw if verificationCode is empty 1`] = `"user id and verification code are required"`;

exports[`#createUserProfileWithDefaultValues should throw an error if email already exists 1`] = `"User already exists"`;

exports[`#createUserProfileWithDefaultValues should throw bad request error if phone number already exists 1`] = `"User already exists"`;

exports[`#getUserAccountWithCredentials should throw if user is not found 1`] = `"Invalid user account"`;

exports[`#getUserProfileById should throw error if user does not exists 1`] = `"Invalid user account"`;
